cf_remote/commands.py

```diff
@@ -679,6 +706,15 @@ def deploy(hubs, masterfiles):
     if os.path.isfile(masterfiles):
         return deploy_tarball(hubs, masterfiles)
 
+    if masterfiles.endswith(".tgz"):
+        if not os.path.exists(masterfiles):
+            log.error("'{}' does not exist".format(masterfiles))
+            return 1
+        else:
+            assert not os.path.isfile(masterfiles)
```

Comment:
I don't understand this section here. Seems like you are saying that if masterfiles exists as a path and is a file, then throw an exception. If it is not a file log an error. So how would masterfiles be used if it is what we expect: a path and a file?

Reply:
Oh. I see, you already handled deploying the tarball above.

```
    if os.path.isfile(masterfiles):
        return deploy_tarball(hubs, masterfiles)
```

Still the assert feels confusing to me. The condition of os.path.isfile(masterfiles) was already checked, handled, and a return was made so this assert is unreachable right?

Reply:
@craigcomstock Even though that if check is _only_ 10 lines above, this can change in the future. Using an assert on something that has already been checked is common / good. It communicates the intentions / assumptions for this code, look at it in combination with the line below:

```
assert not os.path.isfile(masterfiles)
log.error("'{}' is not a file".format(masterfiles))
```

It makes it really clear that at this point in the code, we assume / know that it's a file. Otherwise the error message doesn't make sense. I don't want an if here, because I already know it to be true, so an assert or comment is appropriate. Nobody reads comments, so asserts are better :)

Reply:
yeah, I see your logic but it still reads confusingly for me. Seems like extra work/code for pretty unlikely corner case of a .tgz thing being missing or a directory for a moment and then a file suddenly.